vms/evm/uptimetracker/state.go

```diff
@@ -0,0 +1,291 @@
+		return err
+	}
+
+	s.updatedData[vdr.ValidationID] = updatedStatus
+	return nil
+}
+
+// updateValidator updates the validator in the state
+// returns an error if the validator does not exist or if the immutable fields are modified
+func (s *state) updateValidator(vdr Validator) error {
+	data, ok := s.data[vdr.ValidationID]
+	if !ok {
+		return database.ErrNotFound
+	}
+	// check immutable fields
+	if !data.constantsAreUnmodified(vdr) {
+		return ErrImmutableField
+	}
+	// check if mutable fields have changed
+	updated := deletedStatus
+	if data.IsActive != vdr.IsActive {
+		data.IsActive = vdr.IsActive
+		updated = updatedStatus
+	}
+
+	if data.Weight != vdr.Weight {
+		data.Weight = vdr.Weight
+		updated = updatedStatus
+	}
+
+	s.updatedData[vdr.ValidationID] = updated
+	return nil
+}
+
+// deleteValidator marks the validator as deleted
+// marked validator will be deleted from disk when WriteState is called
+func (s *state) deleteValidator(vID ids.ID) bool {
+	data, ok := s.data[vID]
+	if !ok {
+		return false
+	}
+	delete(s.data, data.validationID)
+	delete(s.index, data.NodeID)
+
+	// mark as deleted for WriteValidator
+	s.updatedData[data.validationID] = deletedStatus
+	return true
+}
+
+// writeState writes the updated state to the disk
+func (s *state) writeState() bool {
+	// TODO: consider adding batch size
+	batch := s.db.NewBatch()
+	for vID, updateStatus := range s.updatedData {
+		switch updateStatus {
+		case updatedStatus:
+			data := s.data[vID]
+
+			dataBytes, err := vdrCodec.Marshal(codecVersion, data)
+			if err != nil {
+				return false
+			}
+			if err := batch.Put(vID[:], dataBytes); err != nil {
+				return false
+			}
+		case deletedStatus:
+			if err := batch.Delete(vID[:]); err != nil {
+				return false
+			}
+		}
+	}
+	if err := batch.Write(); err != nil {
+		return false
+	}
+	// we've successfully flushed the updates, clear the updated marker.
+	clear(s.updatedData)
+	return true
+}
+
+// parseValidatorData parses the data from the bytes into given validatorData
+func parseValidatorData(bytes []byte, data *validatorData) error {
+	if len(bytes) != 0 {
+		if _, err := vdrCodec.Unmarshal(bytes, data); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Load the state from the disk
+func (s *state) loadFromDisk() error {
+	it := s.db.NewIterator()
+	defer it.Release()
+	for it.Next() {
+		vIDBytes := it.Key()
+		vID, err := ids.ToID(vIDBytes)
+		if err != nil {
+			return fmt.Errorf("failed to parse validator ID: %w", err)
+		}
+		vdr := &validatorData{
+			validationID: vID,
+		}
+		if err := parseValidatorData(it.Value(), vdr); err != nil {
+			return fmt.Errorf("failed to parse validator data: %w", err)
+		}
+		if err := s.addData(vID, vdr); err != nil {
+			return err
+		}
+	}
+	return it.Error()
+}
+
+// addData adds the data to the state
+// returns an error if the data already exists
+func (s *state) addData(vID ids.ID, data *validatorData) error {
+	if _, exists := s.data[vID]; exists {
```

Comment:
nit: `ok` naming - we should do a pass over the PR for this since it's used in a lot of places